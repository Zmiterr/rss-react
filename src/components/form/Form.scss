.form {
	max-width: 500px;
	text-align: center;
	border: 2px solid #78DEC7;
	margin: 0 auto;
	box-shadow: 0 0 10px rgb(120 122 199 / 35%);

	&__check {
		margin: 0.3em;
		font-size: 0.7em;
		color: red;
	}

	&__submit {
		margin-top: 25px;
	}

	input,
	select {
		display: block;
		margin: 0 auto;
		margin-bottom: 15px;
		height: 25px;
		width: 150px;
	}

	input[type="radio"] {
		display: inline-block;
		width: 45px;
		position: relative;
		margin: 0;
	}

	.label__radio {
		div {
			display: flex;
			justify-content: center;
			align-items: center;
			margin-top: 10px;
		}

		span {
			width: 30px;
		}
	}
}
