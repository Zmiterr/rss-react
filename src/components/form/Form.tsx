import React, { ReactElement } from 'react';
import { useDispatch } from 'react-redux';
import storeSelector from '../../hooks/useTypedSelector';
import newsApi, { apiKey } from '../../services/api';
import { StatusCode } from '../../types/types';
import { ActionTypes } from '../../types/actions';
import DropDown from '../dropDown';
import PageNumber from '../pageNumber';
import PageSize from '../pageSize';
import SearchPanel from '../search-panel';

import './Form.scss';

const Form = (): ReactElement => {
	const dispatch = useDispatch();
	const { searchValue, sortBy } = storeSelector((state) => state.articles);
	const { page, pageSize } = storeSelector((state) => state.page);

	const setErrorStatus = (status: number) => {
		const obj = {
			NOT_FOUND: 'Data not found. Try something else ..',
			SERV_LATE: 'Server side error. Try later...',
			SERV_DATA: 'Server side error. Enter other data..',
			SERV_LIMIT: 'Limit of server requests. Try later..',
		};
		if (status === StatusCode.ClientErrorBadRequest) {
			dispatch({ type: ActionTypes.GET_DATA_SUCCESS, payload: obj.NOT_FOUND });
		}
		if (status === StatusCode.ClientErrorUpgradeRequired) {
			dispatch({ type: ActionTypes.GET_DATA_SUCCESS, payload: obj.SERV_DATA });
		}
		if (status === StatusCode.ClientErrorTooManyRequests) {
			dispatch({ type: ActionTypes.GET_DATA_SUCCESS, payload: obj.SERV_LATE });
		}
		if (status === StatusCode.ServerErrorInternal) {
			dispatch({ type: ActionTypes.GET_DATA_SUCCESS, payload: obj.SERV_LIMIT });
		}
	};

	const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
		event.preventDefault();
		dispatch({ type: ActionTypes.GET_LOADING_SUCCESS, payload: true });
		try {
			const response = await newsApi.get(
				`v2/everything?q=${searchValue}&sortBy=${sortBy}&pageSize=${pageSize}&page=${page}&apiKey=${apiKey}`
			);
			const { articles, totalResults } = response.data;

			dispatch({ type: ActionTypes.ADD_ARTICLE, payload: articles });
			dispatch({ type: ActionTypes.TOTAL_RESULTS, payload: totalResults });
			dispatch({ type: ActionTypes.GET_DATA_SUCCESS, payload: '' });
		} catch (e: any) {
			// TODO make type for error
			setErrorStatus(e.response.status);
			dispatch({ type: ActionTypes.ADD_ARTICLE, payload: [] });
			setErrorStatus(e.response.status);
			throw new Error(e);
		} finally {
			dispatch({ type: ActionTypes.GET_LOADING_SUCCESS, payload: false });
			dispatch({ type: ActionTypes.PAGE_COUNTER, payload: page });
		}
	};

	return (
		<form className="form" onSubmit={handleSubmit}>
			<SearchPanel searchValue={searchValue} />
			<DropDown sortBy={sortBy} />
			<PageSize pageSize={pageSize} />
			<PageNumber page={page} />
		</form>
	);
};

export default Form;
