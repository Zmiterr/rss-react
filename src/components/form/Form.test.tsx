import React from "react";
import Form from './Form';
//test deps
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { BrowserRouter } from 'react-router-dom';

Enzyme.configure({ adapter: new Adapter() })


describe("Form component", () => {
	it("should render Form component", () => {
		const component = shallow(<BrowserRouter><Form /></BrowserRouter>);
		expect(component).toMatchSnapshot();
	});
});
