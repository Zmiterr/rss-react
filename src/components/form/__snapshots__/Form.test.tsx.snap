// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should render Form component 1`] = `ShallowWrapper {}`;

exports[`Input component should render Input component 1`] = `ShallowWrapper {}`;
